File: ts/packages/luzid-grpc-client/src/client/rpc.ts
import { createClient } from '../connection'
import type { Channel } from '../connection'

import {
  RpcGetAccountInfoRequest,
  RpcGetAccountInfoResponse,
  RpcGetAccountInfoServiceDefinition,
  RpcGetAccountInfoServiceClient,
  RpcRequestAirdropRequest,
  RpcRequestAirdropResponse,
  RpcRequestAirdropServiceDefinition,
  RpcRequestAirdropServiceClient,
} from '@luzid/grpc'

export {
  RpcContext,
  RpcGetAccountInfoRequest,
  RpcGetAccountInfoResponse,
  RpcRequestAirdropRequest,
  RpcRequestAirdropResponse,
} from '@luzid/grpc'

// -----------------
// RpcGetAccountInfoClient
// -----------------
class RpcGetAccountInfoClient {
  private readonly client: RpcGetAccountInfoServiceClient

  constructor(channel: Channel) {
    this.client = createClient(RpcGetAccountInfoServiceDefinition, channel)
  }

  getAccountInfo(
    request: RpcGetAccountInfoRequest
  ): Promise<RpcGetAccountInfoResponse> {
    return this.client.getAccountInfo(request)
  }
}

// -----------------
// RpcRequestAirdropClient
// -----------------
class RpcRequestAirdropClient {
  private readonly client: RpcRequestAirdropServiceClient

  constructor(channel: Channel) {
    this.client = createClient(RpcRequestAirdropServiceDefinition, channel)
  }

  requestAirdrop(
    request: RpcRequestAirdropRequest
  ): Promise<RpcRequestAirdropResponse> {
    return this.client.requestAirdrop(request)
  }
}

// -----------------
// Consolidated RpcClient
// -----------------
export class RpcClient {
  private readonly getAccountInfoClient: RpcGetAccountInfoClient
  private readonly requestAirdropClient: RpcRequestAirdropClient

  constructor(channel: Channel) {
    this.getAccountInfoClient = new RpcGetAccountInfoClient(channel)
    this.requestAirdropClient = new RpcRequestAirdropClient(channel)
  }

  getAccountInfo(
    request: RpcGetAccountInfoRequest
  ): Promise<RpcGetAccountInfoResponse> {
    return this.getAccountInfoClient.getAccountInfo(request)
  }

  requestAirdrop(
    request: RpcRequestAirdropRequest
  ): Promise<RpcRequestAirdropResponse> {
    return this.requestAirdropClient.requestAirdrop(request)
  }
}

File: dart/luzid-grpc-client/lib/src/client/rpc.dart
import 'package:grpc/grpc_connection_interface.dart';
import 'package:luzid_grpc/luzid_grpc.dart';

// -----------------
// RpcGetAccountInfoClient
// -----------------
class RpcGetAccountInfoClient {
  final RpcGetAccountInfoServiceClient _client;

  RpcGetAccountInfoClient(ClientChannelBase channel)
      : _client = RpcGetAccountInfoServiceClient(channel);

  Future<RpcGetAccountInfoResponse> getAccountInfo(
    RpcGetAccountInfoRequest request,
  ) {
    return _client.getAccountInfo(request);
  }
}

// -----------------
// RpcRequestAirdropClient
// -----------------
class RpcRequestAirdropClient {
  final RpcRequestAirdropServiceClient _client;

  RpcRequestAirdropClient(ClientChannelBase channel)
      : _client = RpcRequestAirdropServiceClient(channel);

  Future<RpcRequestAirdropResponse> requestAirdrop(
    RpcRequestAirdropRequest request,
  ) {
    return _client.requestAirdrop(request);
  }
}

// -----------------
// Consolidated RpcClient
// -----------------
class RpcClient {
  final RpcGetAccountInfoClient getAccountInfoClient;
  final RpcRequestAirdropClient requestAirdropClient;

  RpcClient(ClientChannelBase channel)
      : getAccountInfoClient = RpcGetAccountInfoClient(channel),
        requestAirdropClient = RpcRequestAirdropClient(channel);

  Future<RpcGetAccountInfoResponse> getAccountInfo(
    RpcGetAccountInfoRequest request,
  ) {
    return getAccountInfoClient.getAccountInfo(request);
  }

  Future<RpcRequestAirdropResponse> requestAirdrop(
    RpcRequestAirdropRequest request,
  ) {
    return requestAirdropClient.requestAirdrop(request);
  }
}
